src/Vault.ts

```diff
@@ -0,0 +1,419 @@
+    this.pool = network.pool
+    this._asset = asset.toLowerCase() as HexString
+    this.address = address.toLowerCase() as HexString
+  }
+
+  /**
+   * Get the contract address of the restriction mananger.
+   * @internal
+   */
+  _restrictionManager() {
+    return this._query(['restrictionManager'], () =>
+      this.network._share(this.trancheId).pipe(
+        switchMap(
+          (share) =>
+            this._root.getClient(this.chainId)!.readContract({
+              address: share,
+              abi: ABI.Currency,
+              functionName: 'hook',
+            }) as Promise<HexString>
+        )
+      )
+    )
+  }
+
+  /**
+   * Get the details of the investment currency.
+   */
+  investmentCurrency() {
+    return this._root.currency(this._asset, this.chainId)
+  }
+
+  /**
+   * Get the details of the share token.
+   */
+  shareCurrency() {
+    return this.network.shareCurrency(this.trancheId)
+  }
+
+  /**
+   * Get the allowance of the investment currency for the CentrifugeRouter,
+   * which is the contract that moves funds into the vault on behalf of the investor.
+   * @param owner - The address of the owner
+   */
+  allowance(owner: string) {
+    const address = owner.toLowerCase()
+    return this._query(['allowance', address], () =>
+      this.investmentCurrency().pipe(
+        switchMap((currency) =>
+          defer(() =>
+            this._root
+              .getClient(this.chainId)!
+              .readContract({
+                address: this._asset,
+                abi: ABI.Currency,
+                functionName: 'allowance',
+                args: [address, lpConfig[this.chainId]!.centrifugeRouter],
+              })
+              .then((val: any) => new Currency(val, currency.decimals))
+          ).pipe(
+            repeatOnEvents(
+              this._root,
+              {
+                address: this._asset,
+                abi: ABI.Currency,
+                eventName: ['Approval', 'Transfer'],
+                filter: (events) => {
+                  return events.some((event) => {
+                    return (
+                      event.args.owner?.toLowerCase() === address ||
+                      event.args.spender?.toLowerCase() === this._asset ||
+                      event.args.from?.toLowerCase() === address
+                    )
+                  })
+                },
+              },
+              this.chainId
+            )
+          )
+        )
+      )
+    )
+  }
+
+  /**
+   * Get the details of the investment of an investor in the vault and any pending investments or redemptions.
+   * @param investor - The address of the investor
+   */
+  investment(investor: string) {
+    const address = investor.toLowerCase() as HexString
+    return this._query(['investment', address], () =>
+      combineLatest([
+        this.investmentCurrency(),
+        this.shareCurrency(),
+        this.network._investmentManager(),
+        this._restrictionManager(),
```

Comment:
Will this be somehow magically combined into 1 big multicall, or will it be 4 separate calls?

If the latter, would it be worth combining more of these read calls into 1 method?

Reply:
Viem has the option to automatically batch these, which is awesome, so these will be done as a multicall 🥳

Reply:
That is wild! 🔥